### PR TITLE
Remove ExternalPort functions, now Envoy provides new functionality.

### PR DESCRIPTION
**Pull request by youngnick**
Updates #1125 
Updates #210 

This update removes the functionality of the `--envoy-external-http-port` and `--envoy-external-https-port` flags, as we are now able to pass `www.example.com` and `www.example.com:*` to Envoy as allowed `Host:` headers, which makes them not required.

If you need to make your services available on a non-standard load balancer port (ie not 80 for http and 443 for https) upstream of Envoy, that no longer requires any configuration. As long as the traffic arrives at Envoy, it will be routed by the Host header, and will match any external port.

Note that this should also allow the easier use of minikube and kind for Contour testing, another PR will follow with documentation for both use cases.

### CODE REVIEW

File: cmd/contour/contour.go [hunk 1]
```diff
@@ -140,6 +138,13 @@ func main() {
 	serve.Flag("ingress-class-name", "Contour IngressClass name").StringVar(&reh.IngressClass)
 	serve.Flag("ingressroute-root-namespaces", "Restrict contour to searching these namespaces for root ingress routes").StringVar(&ingressrouteRootNamespaceFlag)
 
+	// The following flags are no-ops, and the variables are used to print a message that they don't do anything
+	// any more.
+	var externalHTTPPort int
```
Comment (davecheney) [inline cmd/contour/contour.go]: Please add
```
// TODO(youngnick) remove these for 0.14, see #nnnn 
```
Where `nnnn` is an issue to remove these flags

File: cmd/contour/contour.go [hunk 2]
```diff
@@ -163,6 +168,14 @@ func main() {
 		log.Infof("args: %v", args)
 		var g workgroup.Group
 
+		// Functionality warnings for non-functional flags.
+		if externalHTTPPort != 0 {
+			log.Infof("Due to changes to upstream Envoy, setting envoy-external-http-port to %d no longer does anything, your Ingress will work without it. This notice will be removed in two releases.", externalHTTPPort)
```
Comment (davecheney) [inline cmd/contour/contour.go]: shorter is better

log.Warn("--envoy-external-http-port will be removed in 0.14")

File: cmd/contour/contour.go [hunk 3]
```diff
@@ -163,6 +168,14 @@ func main() {
...
+		}
+		if externalHTTPSPort != 0 {
+			log.Infof("Due to changes to upstream Envoy, setting envoy-external-https-port to %d no longer does anything, your Ingress will work without it. This notice will be removed in two releases.", externalHTTPSPort)
```
Comment (davecheney) [inline cmd/contour/contour.go]: same

File: cmd/contour/contour.go [hunk 4]
```diff
@@ -163,6 +168,14 @@ func main() {
 		log.Infof("args: %v", args)
 		var g workgroup.Group
 
+		// Functionality warnings for non-functional flags.
```
Comment (davecheney) [inline cmd/contour/contour.go]: s/Functionality/Deprecation warning for deprecated flags.

File: internal/contour/route_test.go [hunk 5]
```diff
@@ -237,7 +237,7 @@ func TestRouteVisit(t *testing.T) {
 					Name: "ingress_http",
 					VirtualHosts: []route.VirtualHost{{
 						Name:    "www.example.com",
-						Domains: []string{"www.example.com", "www.example.com:80"},
+						Domains: []string{"www.example.com", "www.example.com:*"},
```
Comment (davecheney) [inline internal/contour/route_test.go]: suggestion: the value of `Domains` can be determined fully from its input, try making a small helper to dry this up.
```go
func domains(hostname string) []string {
 return []string{hostname, hostname+":*"}
}
```

File: internal/dag/builder.go [hunk 6]
```diff
@@ -39,17 +39,8 @@ const (
 type Builder struct {
 	KubernetesCache
 
-	// ExternalInsecurePort is the port that HTTP
-	// requests will arrive at the ELB or NAT that
-	// presents Envoy at the edge network.
-	// If not supplied, defaults to 80.
-	ExternalInsecurePort int
+	// TODO: Remove this layer of indirection?
```
Comment (davecheney) [inline internal/dag/builder.go]: // TODO(youngnick) merge Builder and Kubernetes cache? See #nnnn

Comment (davecheney) [reply internal/dag/builder.go]: The TODO doesn't mean you'll do it, it just means you're the person who knows best what that comment means.

File: internal/dag/builder_test.go [hunk 7]
```diff
@@ -1664,9 +1664,7 @@ func TestDAGInsert(t *testing.T) {
 			),
 		},
 		"insert ingress w/ tls with different secure port": {
```
Comment (davecheney) [inline internal/dag/builder_test.go]: Is this test testing anything anymore?

Comment (youngnick) [reply internal/dag/builder_test.go]: No, and it was broken. Oops, thanks, removed.

File: internal/dag/builder.go [hunk 8]
```diff
@@ -229,7 +220,7 @@ func (b *builder) lookupSecret(m meta, validate func(*v1.Secret) bool) *Secret {
 }
 
 func (b *builder) lookupVirtualHost(name string) *VirtualHost {
-	l := b.listener(b.externalInsecurePort())
+	l := b.listener(80)
```
Comment (stevesloka) [inline internal/dag/builder.go]: Won't this break if we wanted to serve on a different port, say 8080 since the listener is now hardcoded to 80/443?

Comment (youngnick) [reply internal/dag/builder.go]: I thought that too, but this doesn't actually configure the listener at all! It's just a magic number that chooses whether or not the listener in the dag is the secure or insecure one. The actual listening ports are determined purely by the Ingress object. That's why I opened #1135, to clean this up.

Comment (youngnick) [reply internal/dag/builder.go]: I've tested this with Ingress documents that listen on random ports, and it all works just fine.

Comment (stevesloka) [reply internal/dag/builder.go]: :+1:  :tada:

File: internal/dag/builder_test.go [hunk 9]
```diff
@@ -3685,10 +3662,7 @@ func TestDAGIngressRouteUniqueFQDNs(t *testing.T) {
 
 	for name, tc := range tests {
 		t.Run(name, func(t *testing.T) {
-			b := Builder{
-				ExternalInsecurePort: 10080,
-				ExternalSecurePort:   10443,
-			}
+			b := Builder{}
```
Comment (davecheney) [inline internal/dag/builder_test.go]: nit, you can write 
```go
var b Builder
```
This is perhaps a matter of taste but I use the `var` form to indicate to the reader that we're declaring a `Builder` without explicitly initalising it. This leaves the `Builder{ ... }` form for when explicit initialisation of `Builder`'s fields is needed.

File: internal/dag/builder_test.go [hunk 10]
```diff
@@ -1663,29 +1663,6 @@ func TestDAGInsert(t *testing.T) {
 				},
 			),
 		},
-		"insert ingress w/ tls with different secure port": {
```
Comment (davecheney) [inline internal/dag/builder_test.go]: Around line 1375 you'll find
```go
        tests := map[string]struct {
                *Builder
                objs []interface{}
                want []Vertex
        }{
```
now there is no need to provide a custom `*Builder` as part of the fixture -- as there is nothing to customise -- this field can probably be removed and the logic on like 3688 cleaned up.

File: internal/e2e/rds_test.go [hunk 11]
```diff
@@ -309,7 +309,7 @@ func TestEditIngressInPlace(t *testing.T) {
 				Name: "ingress_http",
 				VirtualHosts: []route.VirtualHost{{
 					Name:    "hello.example.com",
-					Domains: []string{"hello.example.com", "hello.example.com:80"},
+					Domains: []string{"hello.example.com", "hello.example.com:*"},
```
Comment (davecheney) [inline internal/e2e/rds_test.go]: perhaps copy the `domains` helper from `contour/route_test.go` and clean this up a little.

File: internal/dag/builder_test.go (hunk 9 above)
Comment (youngnick) [reply internal/dag/builder_test.go]: I read this in one of your blogs, ( 😄 ) will do.

File: internal/dag/builder_test.go (hunk 10 above)
Comment (youngnick) [reply internal/dag/builder_test.go]: Removed the `*Builder` and just made a new Builder in the actual execution loop down below.

File: internal/e2e/rds_test.go (hunk 11 above)
Comment (youngnick) [reply internal/e2e/rds_test.go]: Good idea, done.

**Pull request merged**